File: src/main/java/edu/practice/datastructures/tree/BinarySearchTree.java
package edu.practice.datastructures.tree;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class BinarySearchTree<E extends Comparable<E>> {

    private static class Node<E> {
        Node<E> left;
        E data;
        Node<E> right;

        Node(E data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    private Node<E> root;
    private int size;

    public BinarySearchTree() {
    }

    public int size() { return size; }

    public boolean isEmpty() {
        return (size == 0);
    }

    public E getRoot() {
        if (isEmpty()) {
            return null;
        }

        return root.data;
    }

    private Node<E> addItemToNode(E item, Node<E> currentNode) {

        if (currentNode == null) {
            return new Node<>(item);
        }

        if (item.compareTo(currentNode.data) < 0) {
            currentNode.left =  addItemToNode(item, currentNode.left);
        } else if (item.compareTo(currentNode.data) >= 0) {
            currentNode.right = addItemToNode(item, currentNode.right);
        }

        return currentNode;
    }

    public void add(E item) {
        root = addItemToNode(item, root);

        size++;
    }

    private Node<E> findSmallestNode(Node<E> node) {
        if (node == null) {
            return null;
        }

        if (node.left == null) {
            return node;
        }

        return findSmallestNode(node.left);
    }

    private Node<E> removeNodeWithItem(E item, Node<E> currentNode) {
        if (currentNode == null) { return null; }

        int itemComparedToNodeData = item.compareTo(currentNode.data);

        if (itemComparedToNodeData < 0) {
            currentNode.left = removeNodeWithItem(item, currentNode.left);
        } else if (itemComparedToNodeData > 0) {
            currentNode.right = removeNodeWithItem(item, currentNode.right);
        } else { // Node's data == item to remove

            if (currentNode.left == null && currentNode.right == null) {
                // No children
                return null;
            } else if (currentNode.left == null) {
                // Only right child
                return currentNode.right;
            } else if (currentNode.right == null) {
                // Only right child
                return currentNode.left;
            } else {
                // Both children
                Node<E> newSuccessorNode = findSmallestNode(currentNode.right);
                // Current Node copies data from the new successor's data
                currentNode.data = newSuccessorNode.data;
                // Remove the new successor's node which contains the value to delete
                currentNode.right = removeNodeWithItem(newSuccessorNode.data, currentNode.right);
            }
        }

        return currentNode;
    }

    public void remove(E item) {
        root = removeNodeWithItem(item, root);

        size--;
    }

    private <T> T checkIfItemPresentInNode(E item, Node<E> currentNode, Function<Node<E>, T> action) {
        if (currentNode == null) {
            return action.apply(null);
        }

        int itemComparedToNodeData = item.compareTo(currentNode.data);

        if (itemComparedToNodeData < 0) {
            return checkIfItemPresentInNode(item, currentNode.left, action);
        } else if (itemComparedToNodeData > 0) {
            return checkIfItemPresentInNode(item, currentNode.right, action);
        } else { // Node's data == item to check for
            return action.apply(currentNode);
        }
    }

    public boolean contains(E item) {
        return checkIfItemPresentInNode(item, root, node -> node != null && node.data.equals(item));
    }


    public boolean isLeaf(E item) {
        return checkIfItemPresentInNode(item, root, node ->
                node.left == null && node.right == null
        );
    }

    private int calculateHeightOfNode(Node<E> node) {
        if (node == null) {
            return -1;
        }

        int leftHeight = calculateHeightOfNode(node.left);
        int rightHeight = calculateHeightOfNode(node.right);

        return Math.max(leftHeight, rightHeight) + 1;
    }

    public int height() {
        return calculateHeightOfNode(root);
    }

    private void inorderTraversalToGivenList(List<E> resultList, Node<E> currentNode) {
        if (currentNode == null) { return; }

        inorderTraversalToGivenList(resultList, currentNode.left);

        resultList.add(currentNode.data);

        inorderTraversalToGivenList(resultList, currentNode.right);
    }

    public List<E> inorderTraversal() {
        List<E> result = new ArrayList<>();
        inorderTraversalToGivenList(result, root);
        return result;
    }

    private void preorderTraversalToGivenList(List<E> resultList, Node<E> currentNode) {
        if (currentNode == null) { return; }

        resultList.add(currentNode.data);

        preorderTraversalToGivenList(resultList, currentNode.left);

        preorderTraversalToGivenList(resultList, currentNode.right);
    }

    public List<E> preorderTraversal() {
        List<E> result = new ArrayList<>();
        preorderTraversalToGivenList(result, root);
        return result;
    }

    private void postorderTraversalToGivenList(List<E> resultList, Node<E> currentNode) {
        if (currentNode == null) { return; }

        postorderTraversalToGivenList(resultList, currentNode.left);

        postorderTraversalToGivenList(resultList, currentNode.right);

        resultList.add(currentNode.data);
    }

    public List<E> postorderTraversal() {
        List<E> result = new ArrayList<>();
        postorderTraversalToGivenList(result, root);
        return result;
    }
}

File: src/main/kotlin/datastructures/tree/BinarySearchTree.kt
package edu.practice.datastructures.tree

class BinarySearchTree<E: Comparable<E>>: Tree<E> {

    private data class Node<E>(var data: E, var left: Node<E>? = null, var right: Node<E>? = null)

    private var root: Node<E>? = null
    private var size = 0

    override fun insert(item: E) {
        fun addItemToNode(item: E, currentNode: Node<E>?): Node<E> {
            currentNode?.let {
                when {
                    item < it.data -> it.left = addItemToNode(item, it.left)
                    else -> it.right = addItemToNode(item, it.right)
                }
                return it
            } ?: return Node(item)
        }

        root = addItemToNode(item, root)
        size++
    }

    override fun delete(item: E): Boolean {
        if (root == null) return false

        root = removeNodeWithItem(item, root)
        size--
        return true
    }

    private fun findSmallestNode(node: Node<E>?): Node<E>? {
        return node?.left?.let { findSmallestNode(it) } ?: node
    }

    private fun removeNodeWithItem(item: E, currentNode: Node<E>?): Node<E>? {
        currentNode ?: return null

        when {
            item < currentNode.data -> {
                currentNode.left = removeNodeWithItem(item, currentNode.left)
                return currentNode
            }

            item > currentNode.data -> {
                currentNode.right = removeNodeWithItem(item, currentNode.right)
                return currentNode
            }

            else -> { // Node's data == item to remove
                when {
                    currentNode.left == null && currentNode.right == null -> return null // No children
                    currentNode.left == null -> return currentNode.right // Only Right Child
                    currentNode.right == null -> return currentNode.left // Only Left Child

                    else -> { // Both Children
                        val successorNode = findSmallestNode(currentNode.right)
                        currentNode.data = successorNode?.data ?: return null
                        currentNode.right = removeNodeWithItem(successorNode.data, currentNode.right)
                        return currentNode
                    }
                }
            }
        }
    }

    override fun root(): E? = root?.data

    override fun preOrderTraversal(): List<E> {
        val result: MutableList<E> = mutableListOf()

        fun preorder(currentNode: Node<E>?) {
            if (currentNode == null) return

            result.add(currentNode.data)
            preorder(currentNode.left)
            preorder(currentNode.right)
        }
        preorder(root)

        return result
    }

    override fun inOrderTraversal(): List<E> {
        val result: MutableList<E> = mutableListOf()

        fun inorder(currentNode: Node<E>?) {
            if (currentNode == null) return

            inorder(currentNode.left)
            result.add(currentNode.data)
            inorder(currentNode.right)
        }
        inorder(root)
        println(result)

        return result
    }

    override fun postOrderTraversal(): List<E> {
        val result: MutableList<E> = mutableListOf()

        fun postorder(currentNode: Node<E>?) {
            if (currentNode == null) return

            postorder(currentNode.left)
            postorder(currentNode.right)
            result.add(currentNode.data)
        }
        postorder(root)

        return result
    }

    override fun height(): Int = calculateHeightOfNode(root)

    private fun calculateHeightOfNode(node: Node<E>?): Int {
        if (node == null) { // Base Condition
            return -1
        }

        val leftSubTreeHeight = calculateHeightOfNode(node.left)
        val rightSubTreeHeight = calculateHeightOfNode(node.right)

        return maxOf(leftSubTreeHeight, rightSubTreeHeight) + 1 // 1 for current node
    }

    override fun contains(item: E): Boolean {
        return findNodeAndApply(item, root) {
            node -> node != null && node.data == item
        }
    }

    override fun isLeaf(item: E): Boolean {
        return findNodeAndApply(item, root) {
                node -> node?.left == null && node?.right == null
        }
    }

    private fun <T> findNodeAndApply(item: E, currentNode: Node<E>?, action: (Node<E>?) -> T): T {
        currentNode?.let {
            when {
                item < it.data -> return findNodeAndApply(item, it.left, action)
                item > it.data -> return findNodeAndApply(item, it.right, action)
                else -> return action(it) // Node to apply on
            }
        } ?: return action(null)
    }

    override fun size(): Int = size

    override fun isEmpty(): Boolean = (size == 0)
}